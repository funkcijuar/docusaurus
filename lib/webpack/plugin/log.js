const chalk = require('chalk');
const ora = require('ora');
module.exports = class LogPlugin {
  constructor (options) {
    this.options = options
  }

  apply (compiler) {
    const logger = ora();
    let isFirst = true
    compiler.hooks.done.tap('blogi-log', stats => {
      clearScreen()

      const { port, publicPath } = this.options
      const time = new Date().toTimeString().match(/^[\d:]+/)[0]

      logger.succeed(`${chalk.gray(`[${time}]`)} Build ${chalk.yellow(stats.hash.slice(0, 6))} finished in ${chalk.green(stats.endTime - stats.startTime)} ms!`)
      if (isFirst) {
        isFirst = false
        console.log(`\n${chalk.gray('>')} Blogi dev server listening at ${chalk.cyan(`http://localhost:${port}${publicPath}`)}`)
      }
    })
    compiler.hooks.invalid.tap('blogi-log', clearScreen)
  }
}

function clearScreen () {
  process.stdout.write('\x1Bc')
}