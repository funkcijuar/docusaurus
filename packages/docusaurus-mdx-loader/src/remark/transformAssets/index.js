/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

const visit = require('unist-util-visit');
const path = require('path');
const url = require('url');
const fs = require('fs-extra');

// Needed to throw errors with computer-agnostic path messages
// Absolute paths are too dependant of user FS
function toRelativePath(filePath) {
  return path.relative(process.cwd(), filePath);
}

async function ensureAssetFileExist(assetPath, sourceFilePath) {
  const assetExists = await fs.exists(assetPath);
  if (!assetExists) {
    throw new Error(
      `Asset ${toRelativePath(assetPath)} used in ${toRelativePath(
        sourceFilePath,
      )} not found.`,
    );
  }
}

async function processLinkNode(node, index, parent, {filePath}) {
  if (!node.url) {
    throw new Error(
      `Markdown link url is mandatory. filePath=${toRelativePath(filePath)}`,
    );
  }
  const parsedUrl = url.parse(node.url);
  const assetPath = node.url;
  if (parsedUrl.protocol) {
    // pathname:// is an escape hatch,
    // in case user does not want his assets to be converted to require calls going through webpack loader
    // we don't have to document this for now,
    // it's mostly to make next release less risky (2.0.0-alpha.59)
    if (parsedUrl.protocol === 'pathname:') {
      node.url = node.url.replace('pathname://', '');
    }
    return;
  }
  if (
    assetPath.match(/#|.md|.mdx/) ||
    path.isAbsolute(assetPath) ||
    !path.extname(assetPath) ||
    !assetPath.startsWith('.')
  ) {
    if (!assetPath.startsWith('!')) {
      return;
    }
  }

  const expectedAssetPath = path.join(
    path.dirname(filePath),
    assetPath.replace(/!.*!/, ''),
  );
  await ensureAssetFileExist(expectedAssetPath, filePath);

  node.type = 'jsx';
  node.value = `<a  target="_blank" ${
    assetPath ? `href={require('${assetPath}').default}` : ''
  } ${node.title ? `title={${node.title}}` : ''} >`;
  const {children} = node;
  delete node.children;

  parent.children.splice(index + 1, 0, {
    type: 'paragraph',
    children,
  });

  parent.children.splice(index + 2, 0, {type: 'jsx', value: '</a>'});
}

const plugin = (options) => {
  const transformer = async (root) => {
    const promises = [];
    visit(root, 'link', (node, index, parent) => {
      promises.push(processLinkNode(node, index, parent, options));
    });
    await Promise.all(promises);
  };
  return transformer;
};

module.exports = plugin;
