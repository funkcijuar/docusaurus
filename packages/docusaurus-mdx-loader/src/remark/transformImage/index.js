/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

const visit = require('unist-util-visit');
const path = require('path');
const url = require('url');
const fs = require('fs-extra');

const plugin = (options) => {
  const transformer = (root) => {
    visit(root, 'image', (node) => {
      if (!url.parse(node.url).protocol) {
        if (!path.isAbsolute(node.url)) {
          if (
            !fs.existsSync(path.join(path.dirname(options.filePath), node.url))
          ) {
            throw new Error(
              `Image ${path.join(
                path.dirname(options.filePath),
                node.url,
              )} used in ${options.filePath} not found.`,
            );
          }
          node.type = 'jsx';
          node.value = `<img ${node.alt ? `alt={"${node.alt}"}` : ''} ${
            node.url
              ? `src={require("!url-loader!${
                  node.url.startsWith('./') ? node.url : `./${node.url}`
                }").default}`
              : ''
          } ${node.title ? `title={"${node.title}"}` : ''} />`;
          if (node.url) {
            delete node.url;
          }
          if (node.alt) {
            delete node.alt;
          }
          if (node.title) {
            delete node.title;
          }
        } else if (!fs.existsSync(path.join(options.staticDir, node.url))) {
          throw new Error(
            `Image ${path.join(options.staticDir, node.url)} used in ${
              options.filePath
            } not found.`,
          );
        }
      }
    });
  };

  return transformer;
};

module.exports = plugin;
