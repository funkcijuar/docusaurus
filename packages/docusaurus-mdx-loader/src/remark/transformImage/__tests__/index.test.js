/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {join} from 'path';
import remark from 'remark';
import mdx from 'remark-mdx';
import vfile from 'to-vfile';
import plugin from '../index';
import slug from '../../slug/index';

const processFixture = async (name, options) => {
  const path = join(__dirname, 'fixtures', `${name}.md`);
  const file = await vfile.read(path);
  const result = await remark()
    .use(slug)
    .use(mdx)
    .use(plugin, {...options, filePath: path})
    .process(file);

  return result.toString();
};

test('fail if image donot exists', async () => {
  expect(
    processFixture('fail', {staticDir: join(__dirname, 'fixtures')}),
  ).rejects.toBeInstanceOf(Error);
});

test('transform md images to <img />', async () => {
  const result = await processFixture('img', {
    staticDir: join(__dirname, 'fixtures'),
  });
  expect(result).toMatchSnapshot();
});
