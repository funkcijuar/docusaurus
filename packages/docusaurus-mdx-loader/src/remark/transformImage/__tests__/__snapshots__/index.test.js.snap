// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform md images to <img /> 1`] = `
"![img](https://example.com/img.png)

<img  src={require(\\"!url-loader!./img.png\\").default}  />

<img alt={\\"img\\"} src={require(\\"!url-loader!./img.png\\").default}  />

<img alt={\\"img\\"} src={require(\\"!url-loader!./img.png\\").default} title={\\"Title\\"} /> ![img](/img.png)

## Heading

\`\`\`md
![img](./img.png)
\`\`\`

<img alt={\\"img\\"} src={require(\\"!url-loader!./img.png\\").default}  />
"
`;
